### PR TITLE
fix: remove jsonb_array and fix jsonb data types such that it can be used in schema design and graphql api

### PR DESCRIPTION
**Pull request by mswertz**
closes https://github.com/molgenis/molgenis-emx2/issues/4405

todo:
- remove any jsonb_array function
- add jsonb to schema design and graphql (it clearly never has worked in practice)

out of scope:
- because jsonb has never worked in graphql we assume no migrations are needed.

### CODE REVIEW

**Review by esthervanenckevort — DISMISSED** [review]
Code looks fine, but is missing a migration step. I would expect a JSONB_ARRAY to be converted to a JSON Object containing an array of JSON objects.

**Comment by mswertz** [comment]
> Code looks fine, but is missing a migration step. I would expect a JSONB_ARRAY to be converted to a JSON Object containing an array of JSON objects.

Not needed, was not useable.

**Comment by svandenhoek** [comment]
If #4323 ends up getting merged before this PR, be sure to remove the following lines after merging with master:
https://github.com/molgenis/molgenis-emx2/blob/23d8cb9f1e7deaa06ce2e3d1815a35c849634355/backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/ColumnTypeRdfMapper.java#L67

https://github.com/molgenis/molgenis-emx2/blob/23d8cb9f1e7deaa06ce2e3d1815a35c849634355/backend/molgenis-emx2-rdf/src/test/java/org/molgenis/emx2/rdf/ColumnTypeRdfMapperTest.java#L151-L152

https://github.com/molgenis/molgenis-emx2/blob/23d8cb9f1e7deaa06ce2e3d1815a35c849634355/backend/molgenis-emx2-rdf/src/test/java/org/molgenis/emx2/rdf/ColumnTypeRdfMapperTest.java#L372-L373

#4323 contains a test that validates whether each `ColumnType` has a mapper to RDF, so pre-emptively removing `JSON_ARRAY` from `src` would cause a failing test and removing test code with source code still present seems illogical.

**Review by svandenhoek — CHANGES_REQUESTED** [review]
- docs -> `use_schema.md`
  - remove `jsonb_array`
  - rename `jsonb` to `json`
- Consider front-end json validation, such as:
  - try-except with `JSON.parse(json);`
  - a regex like https://stackoverflow.com/a/32155765 (link needs conversion from PHP though)
- While during row-creation invalid data is not accepted, if editing an existing row this can still be done. See f.e. https://preview-emx2-pr-4409.dev.molgenis.org/json_input_test/tables/#/Test :

| input | expected | actual |
| --- | --- | --- |
| `{"key":"value"};alert("test")` | error (not valid json) | `"{\"key\":\"value\"};alert(\"test\")"` |
| `"{"key":"value"};alert("test")"` | error (not valid json) | `"\"{\"key\":\"value\"};alert(\"test\")\""` |
| `{}alert("test")` | error (not valid json) | `"{}alert(\"test\")"` |
| `"alert("test")` | error (not valid json) | `"\"alert(\"test\")"` |

- Consider adding tests that include primitives as well (`""`, `"string"`, `-1`, `true`, `null`), see also https://datatracker.ietf.org/doc/html/rfc7159#section-3
  - `null` is converted to an empty field (should stay `null`)
- `{"key":"value1", "key":"value2"}` yields `{"key":"value2"}`. While technically correct, an error or warning would be preferred.

**Comment by mswertz** [comment]
> @mswertz: Normaliter zou ik voorstander zijn van de RFC standaarden zo goed mogelijk te volgen, al is er hier wel een punt te maken dat qua usecase we zelf al datatypes hebben voor string/int/boolean (behalve `null`?) waardoor we liever willen dat een gebruiker dat gebruikt i.p.v. een JSON met enkel 1 primitive. Stel wel voor dat we dan:
> 
> * Dit helder in de documentatie zetten:
> 
> ```
> * json : validates json format (must be an array or object!)
> ```
> 
> * Bij back-end checken of de root node een array/object is, bijvoorbeeld:
> 
> ```java
> JsonNode tree = objectMapper.readTree(jsonString);
>     if(!tree.isArray() && !tree.isObject()) {
>         throw new MolgenisException("JSON is not an array/object.");
>     }
> ```

yes, kun jij dat doen of wil je dat ik het doe?

File: apps/molgenis-components/src/components/forms/formUtils/formUtils.ts
Comment (mswertz) [inline apps/molgenis-components/src/components/forms/formUtils/formUtils.ts:248]: doesn't check if it is an array?

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/utils/MolgenisObjectMapper.java
Comment (mswertz) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/utils/MolgenisObjectMapper.java:63]: nice

**Review by mswertz — COMMENTED** [review]
looks good, one question left. I cannot approve so if you agree with comment can you then approve yourself?

File: apps/molgenis-components/src/components/forms/formUtils/formUtils.ts
Comment (svandenhoek) [reply apps/molgenis-components/src/components/forms/formUtils/formUtils.ts:248]: It should allow arrays:
Scherm­afbeelding 2024-11-26 om 13 56 42

While `null` is also an `object`, it does fail in the form (probably due to `if(!parsed)`):
Scherm­afbeelding 2024-11-26 om 13 58 50

Comment (svandenhoek) [reply apps/molgenis-components/src/components/forms/formUtils/formUtils.ts:248]: I'll have a quick look if I can fix `null` giving the correct error message though (no array/object instead of invalid JSON).

**Review by chinook25 — CHANGES_REQUESTED** [review]
looks good, just some small thingies.

File: apps/molgenis-components/src/components/forms/formUtils/formUtils.ts
Comment (svandenhoek) [reply apps/molgenis-components/src/components/forms/formUtils/formUtils.ts:240]: Same. Apparently it's a bug that originated from the very first version of Javascript but fixing it is deemed as breaking too many things: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof#typeof_null

File: apps/molgenis-components/src/components/forms/formUtils/formUtils.ts
Comment (svandenhoek) [reply apps/molgenis-components/src/components/forms/formUtils/formUtils.ts]: Test code for front-end now also added: https://github.com/molgenis/molgenis-emx2/blob/ff6cdbfac2bf41c497ba66328c93f3f56e1eb33a/apps/molgenis-components/src/components/forms/formUtils/formUtils.test.ts#L380-L391

**Review by svandenhoek — APPROVED** [review]

**Review by chinook25 — APPROVED** [review]

**Comment by svandenhoek** [comment]
@mswertz: I approved and @chinook25 did another review and that feedback is included as well.

**Pull request merged**
